src/Storages/Streaming/StreamingStoreSourceBase.cpp

```diff
@@ -161,13 +168,14 @@ Chunk StreamingStoreSourceBase::generate()
 
 /// It basically initiate a checkpoint
 /// Since the checkpoint method is called in a different thread (CheckpointCoordinator)
-/// We nee make sure it is thread safe
+/// We need make sure it is thread safe
 void StreamingStoreSourceBase::checkpoint(CheckpointContextPtr ckpt_ctx_)
 {
     /// We assume the previous ckpt is already done
     /// Use std::atomic<std::shared_ptr<CheckpointContext>>
-    assert(!ckpt_ctx.load(std::memory_order_relaxed));
-    ckpt_ctx = new CheckpointContext(*ckpt_ctx_);
+    std::scoped_lock lock(ckpt_mutex);
+    assert(!ckpt_ctx);
+    ckpt_ctx = std::make_shared<CheckpointContext>(*ckpt_ctx_);
```

Comment:
I want to know why the previous one was wrong?

Reply:
https://github.com/timeplus-io/proton/issues/320#issuecomment-1821969630
one thread alloc/write by `new`
one thread is access/read with `doCheckpoint`

Reply:
It's an atomic variable, so why does this problem arise?

Reply:
I try many times in my local with my assumption, but do not get a thread sanitizer report(hahaha).

basically, I believe the problem is `ckpt_ctx = new CheckpointContext(*ckpt_ctx_);`
this phase has done 2 things:
the first step is new a object, then the second step will be atomically assigned.

the key is: thread 2(t2) may read old_obj, in the same time ckpt_ctx has been assigning with new_obj in thread 1(t1)
the pointer is safe, but the pointer reference object is not guaranteed. 
with a mutex lock guard, the  t1's new and assign operations are seen as a single, atomic action from t2's perspective.

```
// from gpt
[Thread 1 - Writer]                              [Thread 2 - Reader]
   |                                                   |
   |--- (1) new_obj = new CheckpointContext(...) -----|  // Creates a new object (deep copy)
   |                                                   |
   |--- (2) ckpt_ctx = new_obj -----------------------|  // Atomically assigns the new object's
   |        (Atomic assignment)                        |  // address to ckpt_ctx.
   |                                                   |
   |===================== DATA RACE ===================|
   |                                                   |
   |                                                   |--- (3) current = ckpt_ctx.exchange(nullptr)
   |                                                   |     (Atomic exchange)
   |                                                   |
   |--- (4) Return from function ----------------------|  // Thread 1 completes its operation.
   |                                                   |     // Thread 2 may now be working with
   |                                                   |     // the object previously pointed to by
   |                                                   |     // ckpt_ctx.

```